### PR TITLE
 fix(workflow): comply with branch protection by using PR-based tag updates

### PR DESCRIPTION
**Pull request by KillianGolds**
## Description

This PR updates the release automation workflow to be fully idempotent and improves the release process with enhanced safety, visibility, and reliability features. [RHOAIENG-31952](https://issues.redhat.com/browse/RHOAIENG-31952)

**Key Changes:**

1. **Idempotent Workflow Design** [Example](https://github.com/KillianGolds/odh-model-controller/actions/runs/19679689879)
   - Added tag existence checks to prevent duplicate tag creation
   - Added release existence checks to skip duplicate releases
   - Implemented branch reuse logic for existing branches
   - Added PR existence detection to reuse existing PRs instead of creating duplicates
   - File change detection for graceful no-op behavior when files already have target version

2. **Enhanced Workflow Summary** [Example](https://github.com/KillianGolds/odh-model-controller/actions/runs/19679642752)
   - Added comprehensive workflow summaries showing all actions taken
   - Four summary states: Tag Mismatch Error, No Changes Detected, Dry Run Preview, and Workflow Completed
   - Clickable links to all created resources (tags, releases, PRs, file changes)
   - File diff preview in dry-run mode to review exact changes before execution

3. **Dry-Run Mode Improvements** [Example](https://github.com/KillianGolds/odh-model-controller/actions/runs/19679589071)
   - Dry-run now validates all steps without creating actual tags/releases/PRs
   - Shows complete preview of expected actions and PR details
   - Displays diff previews for both `config/base/params.env` and `.tekton/odh-model-controller-push.yaml`
   - Release existence check runs in all modes for better visibility
   - Enables sanity checks before running the workflow for real

4. **Branch Protection Compliance**
   - Changed from direct pushes to PR-based workflow for Tekton file updates
   - Workflow now creates a PR against the target branch instead of pushing directly
   - Ensures compliance with repository branch protection rules

5. **Early Tag Validation** [Example](https://github.com/KillianGolds/odh-model-controller/actions/runs/19679532631)
   - Validates that `tag_name` exists in the Tekton file before proceeding
   - Fails fast with clear error summary when tags don't match
   - Prevents orphan tag/release creation when wrong tag is specified
   - Shows the actual current tag in file and suggests the correct value

6. **Better Visibility** [Example](https://github.com/KillianGolds/odh-model-controller/actions/runs/19679642752)
   - Extract and display PR numbers for easy reference
   - Direct links to view PR changes, releases, and tags
   - Clear indication of which files were updated and tag progression
   - Graceful error messages when mismatched tags are provided

7. **Code Quality Improvements**
   - Removed unused `NEW_TAG` environment variable
   - Consolidated tag creation flow into 3 clear steps for better maintainability
   - Simplified conditional logic by removing redundant dry-run specific steps
   - Replaced Python tag check with bash script for consistency
   - Replaced git commands with GitHub CLI (`gh`) for branch and tag operations

**Benefits:**
- Safe re-runs without creating duplicate artifacts
- Compliance with branch protection requirements
- Safer releases with dry-run validation
- Improved developer experience with comprehensive summaries
- Early issue detection through preview mode
- Handles edge cases gracefully (mismatched tags, existing artifacts)
- Cleaner, more maintainable workflow code

## How Has This Been Tested?

**Testing Environment:**
- Tested on personal fork [KillianGolds/odh-model-controller](https://github.com/KillianGolds/odh-model-controller/tree/release-test-odh-v3.1)
- Used a mock release branch (`release-test-odh-v3.1`) with both `params.env` and Tekton file set to `odh-v3.1` to simulate real release branch conditions
- Full E2E test suite with 4 comprehensive scenarios found [here](https://github.com/KillianGolds/odh-model-controller/actions/workflows/create-release.yaml).

**Test Scenarios:**

1. **Edge Case: Non-Existent Tag Handling**
   - Ran workflow with tags that don't exist in Tekton file (`odh-v99.1-test` → `odh-v99.2-test`)
   - Verified workflow fails fast with clear "Tag Mismatch" error summary
   - Confirmed no tags, releases, or PRs were created
   - Summary shows the actual current tag and suggests the correct value
   - Example: https://github.com/KillianGolds/odh-model-controller/actions/runs/19679532631

2. **Dry-Run Mode Testing**
   - Ran workflow with dry-run enabled using correct tags (`odh-v3.1` → `odh-v3.2`)
   - Verified no tags, releases, or PRs were created
   - Confirmed workflow summary showed complete preview of expected actions
   - Validated diff display for both `config/base/params.env` and `.tekton/odh-model-controller-push.yaml`
   - Verified release existence check runs in dry-run mode
   - Example: https://github.com/KillianGolds/odh-model-controller/actions/runs/19679589071

3. **End-to-End Workflow Testing (Fresh Run)**
   - Ran workflow without dry-run mode
   - Verified tag and release creation
   - Confirmed PR was created successfully with correct title and body
   - Validated PR updates both `.tekton/odh-model-controller-push.yaml` and `config/base/params.env`
   - Validated all clickable links in workflow summary (tag, release, PR, file changes)
   - Example: https://github.com/KillianGolds/odh-model-controller/actions/runs/19679642752

4. **Idempotency Testing (Re-run)**
   - Re-ran workflow with identical inputs from Test 3
   - Verified workflow detected existing tag and skipped creation
   - Confirmed workflow detected existing release and skipped creation
   - Validated branch reuse logic worked correctly
   - Verified file change detection showed "No Changes Detected" with clear message
   - Confirmed graceful exit with no errors or duplicate artifacts
   - Example: https://github.com/KillianGolds/odh-model-controller/actions/runs/19679689879

All tests passed successfully with expected behavior. Workflow is fully idempotent and handles all edge cases gracefully.

## Merge criteria:

- [x] JIRA(s) are linked in the PR description
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: .github/workflows/create-release.yaml [hunk 1]
```diff
@@ -111,43 +123,171 @@ jobs:
...
+
+          # Ensure next tag is different from current tag
+          if [[ "${{ github.event.inputs.next_odh_tag }}" == "${{ github.event.inputs.tag_name }}" ]]; then
+            echo "::error::Next tag must be different from current tag"
+            exit 1
+          fi
+
+          echo "::notice::Tag validation passed"
+
+      - name: Create branch for ODH tag update
```
Comment (brettmthompson) [inline .github/workflows/create-release.yaml]: It would be good to cleanup any created branches on workflow failures.

Comment (KillianGolds) [reply .github/workflows/create-release.yaml]: This comment here is relevant to this discussion I believe: [HERE](https://github.com/opendatahub-io/odh-model-controller/pull/575/files/c715f8c4ffaea00a2675ca12f513cd49bbcfdc5f#r2454819835)

File: .github/workflows/create-release.yaml [hunk 2]
```diff
@@ -1,153 +1,351 @@
...
+      - name: Get latest tag
+        id: get_tag
+        run: |
+          echo "old_tag_name=$(git ls-remote --tags origin | awk -F'/' '{print $3}' | grep -v '{}' | sort -V | tail -n1)" >> $GITHUB_OUTPUT
+
+      - name: print tag
+        id: print_tag
+        run: |
+          echo "Old Tag=${{ steps.get_tag.outputs.old_tag_name }}"
+          echo "NEW_TAG=${{ github.event.inputs.tag_name }}" >> $GITHUB_ENV
```
Comment (brettmthompson) [inline .github/workflows/create-release.yaml]: Are we using this NEW_TAG value set in the github env anywhere?

Comment (KillianGolds) [reply .github/workflows/create-release.yaml]: Resolved in: [8f6956c](https://github.com/KillianGolds/odh-model-controller/commit/8f6956c1c4ee9df2a7b353bb3f63aa3aada46ba0)

File: .github/workflows/create-release.yaml [hunk 3]
```diff
@@ -1,153 +1,345 @@
...
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ github.ref }}
+
+      - name: Check if release exists
+        id: check_release
+        run: |
+          if gh release view "${{ github.event.inputs.tag_name }}" >/dev/null 2>&1; then
+            echo "release_exists=true" >> $GITHUB_OUTPUT
+            echo "::warning::Release already exists for tag ${{ github.event.inputs.tag_name }}, skipping creation"
```
Comment (spolti) [inline .github/workflows/create-release.yaml]: shouldn't we abort if the release already exists?

Also I think we should clean anything that was created so far (if any)

File: .github/workflows/create-release.yaml [hunk 4]
```diff
@@ -1,153 +1,345 @@
...
+            PR_URL=$(echo "$EXISTING_PR" | jq -r '.url')
+            echo "::notice::PR #$PR_NUMBER already exists for branch $BRANCH_NAME: $PR_URL"
+            echo "pr_url=$PR_URL" >> $GITHUB_OUTPUT
+            echo "pr_number=$PR_NUMBER" >> $GITHUB_OUTPUT
+          else
+            # Create PR with multi-line body
+            PR_BODY="This PR updates the ODH release tag in the Konflux build configuration from \`${{ github.event.inputs.tag_name }}\` to \`${{ github.event.inputs.next_odh_tag }}\` for the next development cycle.
+
+            **Changes:**
+            - Updated \`.tekton/odh-model-controller-push.yaml\` with the new ODH tag
```
[the remark this answers is not in the document]
Comment (KillianGolds) [reply .github/workflows/create-release.yaml]: Thanks for the question @spolti! Looking at the workflow, the params.env update happens in the first job (fetch-tag) at lines 98-111. It updates the image reference from `:fast` to the release tag (e.g., `:odh-v2.35`) and commits that change as part of creating the tag itself.

The bump-odh-tag job (which creates the PR) only updates `.tekton/odh-model-controller-push.yaml` with the next development tag for the Konflux builds.

I could be misunderstanding the intent though - do you think we should also be updating params.env in the PR to point to the next tag as well?

File: .github/workflows/create-release.yaml (hunk 3 above)
Comment (spolti) [reply .github/workflows/create-release.yaml]: I am more concerned about possible garbage left behind by failures.

**Review by mholder6 — APPROVED** [review]
/lgtm
